skema/program_analysis/CAST/matlab/matlab_to_cast.py

```diff
@@ -578,73 +569,119 @@ def visit_do_loop_statement(self, node) -> Loop:
-        mi = self.visit_else_clause(node)
-        # addd comparison operator
-        comp: Operator = get_first_child_by_type(node, "comparison_operator")
-        mi.expr = self.visit(comp)
-
-        return mi
+    def visit_switch_statement(self, node):
+        """ return a conditional statement based on the switch statement """
     
+        # node types used for case comparison
+        case_node_types = self.literal_types + ["identifier"]
+        
+        def get_node_value(ast_node):
+            """ return the CAST node value or var name """
+            if isinstance(ast_node, Var):
+                return ast_node.val.name
+            return ast_node.value
+
+        def get_operator(op, operands):
+            """ return an Operator representing the case test """
+            return Operator(
+                source_language = "matlab",
+                interpreter = None,
+                version = MATLAB_VERSION,
+                op = op,
+                operands = operands
+            )
 
-    def visit_else_clause(self, node):
-        """ Return a ModelIf with body nodes only. """
-        # get the top level body nodes
-        mi = ModelIf()
-        block = get_first_child_by_type(node, "block")
-        for child in block.children:
-            body_node = self.visit(child)
-            if body_node:
-                if not mi.body:
-                    mi.body = list()
-                mi.body.append(body_node)
-
-        return mi
-
+        def get_case_expression(case_node, identifier):
+            """ return an Operator representing the case test """
+            cell_node = get_first_child_by_type(case_node, "cell")
+            # multiple case arguments
+            if (cell_node):
+                nodes = get_all(cell_node, case_node_types)
+                ast_nodes = valid([self.visit(node) for node in nodes])
+                operand = LiteralValue(
+                    value_type="List",
+                    value = [get_node_value(node) for node in ast_nodes],
+                    source_code_data_type=["matlab", MATLAB_VERSION, "unknown"],
+                    source_refs=[self.node_helper.get_source_ref(cell_node)]
+                )
+                return get_operator("in", [identifier, operand])
+            # single case argument
+            nodes = get_children_by_types(case_node, case_node_types)
+            operand = valid([self.visit(node) for node in nodes])[0]
+            return get_operator("==", [identifier, operand])
+
+        def get_case_body(case_node):
+            """ return the instruction block for the case """
+            block = get_first_child_by_type(case_node, "block")
+            if block:
+                return valid([self.visit(child) for child in block.children])
+            return None
+            
+        def get_model_if(case_node, identifier):
+            """ return conditional logic representing the case """
+            return ModelIf(
```

Comment:
Here, you should add the source_ref field that is missing.

Reply:
Fixed